polygon-zkevm-adaptor/tests/end_to_end.rs

```diff
@@ -276,6 +335,180 @@ async fn test_preconfirmations() {
+    tracing::info!("connecting to sequencer at {}", env.sequencer());
+    let sequencer = surf_disco::Client::<hotshot_query_service::Error>::new(env.sequencer());
+    sequencer.connect(None).await;
+
+    // Wait for the adaptor to start serving.
+    tracing::info!("connecting to adaptor RPC at {}", env.l2_adaptor_rpc());
+    // The adaptor is not a full RPC, therefore we can't use `wait_for_rpc`.`
+    wait_for_http(&env.l2_adaptor_rpc(), Duration::from_secs(1), 100)
+        .await
+        .unwrap();
+    tracing::info!(
+        "connecting to adaptor query service at {}",
+        env.l2_adaptor_query()
+    );
+    wait_for_http(&env.l2_adaptor_query(), Duration::from_secs(1), 100)
+        .await
+        .unwrap();
+
+    // Create a test transaction.
+    let transfer_amount = 1.into();
+    let txn_hash = l2
+        .send_transaction(
+            TransactionRequest {
+                from: Some(l2.inner().address()),
+                to: Some(Address::zero().into()),
+                value: Some(transfer_amount),
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .unwrap()
+        .tx_hash();
+    let submitted = Instant::now();
+    tracing::info!("Sent transaction {txn_hash:?} at {submitted:?}");
+
+    // Wait for the transaction to complete on L2, using both the regular RPC and the
+    // preconfirmation RPC in parallel.
+    futures::join!(
+        await_transaction(&l2_preconf, txn_hash),
+        await_transaction(&l2, txn_hash),
+    );
+
+    // Check the effects of the transfer.
+    assert_eq!(
+        l2.get_balance(l2.inner().address(), None).await.unwrap(),
+        l2_initial_balance - transfer_amount
+    );
+    assert_eq!(
+        l2_preconf
+            .get_balance(l2_preconf.inner().address(), None)
+            .await
+            .unwrap(),
+        l2_initial_balance - transfer_amount
+    );
+
+    // Wait for some batches to be verified. The zkevm node does not updated its latest synced L1
+    // block until it reaches a block with relevant events, and for the preconfirmations node, the
+    // only relevant events are from verified batches (since it reads sequenced batches from
+    // HotShot, not the L1). This means that if the preconfirmations node has not seen any verified
+    // batches yet, it has not synced any L1 blocks at all, and the reorg will not affect it.
+    let verified_filter = rollup.verify_batches_trusted_aggregator_filter();
+    verified_filter.stream().await.unwrap().next().await;
+
+    // Wait a few seconds for the preconfirmations node to handle that event.
+    tracing::info!("waiting for batches to be verified");
+    sleep(Duration::from_secs(5)).await;
+
+    // Force an L1 reorg.
+    tracing::info!("causing L1 reorg");
+    reorgme.join();
+
+    // Wait a bit for the nodes to recover.
+    sleep(Duration::from_secs(10)).await;
```

Comment:
Is there some way to check that the nodes have recovered without sleeping? Like some transaction effect we know for sure will have been reversed by the reorg?

Reply:
Ah that comment is a bit misleading. The actual intention of delaying here is to ensure that the nodes have *seen* the reorg, and so that the next transaction we submit is observed after the nodes have observed the reorg. Then the success of that transaction is what tells us the nodes have recovered. 

But looking back at this, I think this was more necessary in my earlier attempts to simulate reorgs, where things were messier and not as determinstic. Now, `reorgme.join()` already waits until it has actually seen the reorg happen on the main L1 RPC node, so this delay might not be necessary at all. I'll see if I can delete it and still get the behavior I want in the logs.

Reply:
Yep, removing this seems to work